### PR TITLE
[unitaryHACK] Add reset error to noise channels

### PR DESCRIPTION
**Pull request by nahumsa**
**Context:** Add Reset Error to noise channels

**Description of the Change:**  Add Single-qubit Reset error channel to pennylane.

**Benefits:** Add a new noise model that simulates the reset of qubits to 0 or 1 given a probability.

**Possible Drawbacks:**

**Related GitHub Issues:** #971

### CODE REVIEW

**Comment by josh146** [comment]
Hi @nahumsa! Thanks for this PR -- this is a great issue to tackle. Just a bit of housekeeping before we can provide a review:

- The formatting check is failing. All that is required is that the PennyLane source code and tests folder is 'blacked' using the Black formatting tool:

  ```python
  pip install black
  black -l 100 pennylane
  black -l 100 tests
  ```

- The documentation check is also failing, due to a missing space after the backtick in the docstring:
  ```
  :math:`p_1 \in [0, 1]`is
  ```

  You can click on the 'details' link beside the `docs/readthedocs.org:pennylane` CI check to view what the rendered documentation for your PR looks like, to make sure it looks okay when built :)

  image

- Finally, it looks like there is an issue with the new tests, causing the test CI to fail.

Let us know when these have been addressed, and the PR is ready for review! Alternatively, feel free to ask any questions.

**Comment by nahumsa** [comment]
Thanks for the response, @josh146 ! I will fix those issues and let you know when it's ready for a review.

**Comment by nahumsa** [comment]
I think that this is ready for review.

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 <h3>New features since last release</h3>
 
+* Added Reset error to noise channels.
```
Comment (josh146) [inline .github/CHANGELOG.md:5]: It might be better here to provide the name of the new operation, so that users can find it easily 🙂 
```suggestion
* Added a new noise channel, `qml.ResetError`.
```

Comment (josh146) [reply .github/CHANGELOG.md:5]: In addition @nahumsa -- feel free to add an example of it's usage here (check out some of the other changelog entries to see how the examples are formatted).

File: pennylane/ops/channel.py [hunk 2]
```diff
@@ -267,6 +267,58 @@ def _kraus_matrices(cls, *params):
...
+    grad_method = "A"
+    grad_recipe = ([[1, 0, 1], [-1, 0, 0]],)
+
+    @classmethod
+    def _kraus_matrices(cls, *params):
+        p_0, p_1 = params[0], params[1]
+        K0 = np.sqrt(1 - p_0 - p_1) * np.eye(2)
+        K1 = np.sqrt(p_0) * np.array([[1, 0], [0, 0]])
+        K2 = np.sqrt(p_1) * np.array([[0, 0], [0, 1]])
+        return [K0, K1, K2]
```
Comment (josh146) [inline pennylane/ops/channel.py:319]: Nice!

File: tests/ops/test_channel_ops.py [hunk 3]
```diff
@@ -213,6 +216,42 @@ def circuit(p):
...
+
+        @qml.qnode(dev)
+        def circuit(p_0, p_1):
+            qml.RX(angle, wires=0)
+            qml.ResetError(p_0, p_1, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        gradient = np.squeeze(qml.grad(circuit)(p_0, p_1))
+        assert gradient == circuit(1) - circuit(0)
+        assert np.allclose(gradient, (-2 * np.cos(angle)))
```
Comment (josh146) [inline tests/ops/test_channel_ops.py:252]: Great test @nahumsa! Did you calculate the expected gradient value, `-2 cos(angle)`, by hand?

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks for this great PR @nahumsa! In particular, fantastic tests. 

However, it doesn't appear that the tests are passing; they are giving the error

```python
AssertionError: Gradient recipe must have one entry for each parameter!
```

The `grad_recipe` you have defined provides a gradient recipe only for the first parameter:

```python
grad_recipe = ([[1, 0, 1], [-1, 0, 0]],)
```

You will need to also provide the gradient recipe for the second parameter, `p_1`. Let me know if you have any questions!

**Comment by josh146** [comment]
Oh, and don't forget to add your name to the Contributors section of the changelog!

**Comment by nahumsa** [comment]
> Oh, and don't forget to add your name to the Contributors section of the changelog!

Sure! :)

File: tests/ops/test_channel_ops.py (hunk 3 above)
Comment (nahumsa) [reply tests/ops/test_channel_ops.py:252]: Yeah, I'm not sure about the analytical gradients for this one, I was using `TestBitFlip` as a reference. 
Maybe could you clarify a bit more about the gradient of this channel?

Comment (josh146) [reply tests/ops/test_channel_ops.py:252]: One of the best sources for learning more about the analytic gradients of noisy channels is https://johannesjakobmeyer.com/blog/004-noisy-parameter-shift/, or alternatively, the PennyLane demonstration: https://pennylane.ai/qml/demos/tutorial_noisy_circuit_optimization.html

Comment (josh146) [reply tests/ops/test_channel_ops.py:252]: In this particular case, I think @ixfoduap calculated the analytic value by hand, so hopefully he will be able to shed some light!

Comment (nahumsa) [reply tests/ops/test_channel_ops.py:252]: Thanks for the followup. I'll read those.

Comment (ixfoduap) [reply tests/ops/test_channel_ops.py:252]: Hi @nahumsa, this is a great PR!

I'm not sure about the analytical gradients. The way we calculate analytical gradients is using this formula, which also appears in the blog post that Josh linked to:
Screenshot from 2021-05-18 14-45-22

It is valid when the channel can be expressed as a linear combination of channels. This typically means that the Kraus matrices themselves should be a valid channel, which in practice often boils down to the Kraus matrices being unitaries.

In your case, the projectors `|0><0|` and `|1><1`| are not by themselves valid channels, so it's not accurate that the reset channel can be expressed as a linear combination of channels. This means that the analytical formula should not apply here. 

I believe that the reason the tests are passing is because you are setting both `p_0` and `p_1` equal to each other, which effectively makes this the identity channel --> `sqrt(p) |0><0| + sqrt(p) |1><1| = sqrt(p) (|0><0|+|1><1|) = sqrt(p) \identity` and we end up with the gradient of the gate in the circuit.

I would expect the test to fail if you make `p_0 != p_1` and compute the gradient by hand there. It's worth checking that this is the case. In any case, the gradient test should be checking gradients for many values of `p_0` and `p_1`.

If the analytical gradients do not apply here, it is not a big deal; it just means you should change the gradient method to "F" (finite difference) like it is done for other channels.

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: tests/ops/test_channel_ops.py (hunk 3 above)
Comment (nahumsa) [reply tests/ops/test_channel_ops.py:252]: Hi @ixfoduap! Thanks for the clarification. After reading the reference, I think that it is not possible to get analytic gradient by the formulas used on other channels.  
Thanks for the catch about the tests. After changing the gradient for using the finite difference methods should I delete the gradient test?

Comment (josh146) [reply tests/ops/test_channel_ops.py]: Thanks @ixfoduap!

Comment (josh146) [reply tests/ops/test_channel_ops.py]: @nahumsa let us know when you have the tests passing, or if you need a hand anywhere :)

Comment (nahumsa) [reply tests/ops/test_channel_ops.py]: @josh146 and @ixfoduap, after revising some math, I found that I missed two Kraus operators that constitutes the `ResetError` channel! Sorry about that, I always find quantum channels quite tricky 😅

**Review by josh146 — APPROVED** [review]
Great contribution @nahumsa, looks ready to merge on my end! I might wait to see if @ixfoduap has any requests.

File: pennylane/ops/channel.py [hunk 4]
```diff
@@ -267,6 +267,71 @@ def _kraus_matrices(cls, *params):
...
+                \end{bmatrix}
+
+    .. math::
+        K_1 = \sqrt{p_0}\begin{bmatrix}
+                1 & 0  \\
+                0 & 0
+                \end{bmatrix}
+
+    .. math::
+        K_2 = \sqrt{p_0}\begin{bmatrix}
```
Comment (ixfoduap) [inline pennylane/ops/channel.py:289]: @nahumsa Why did you change the definition of the channel? 
Before approving, could you please point me to a reference where this channel is defined? I would like to make sure that the name "reset error" coincides with the standard usage. Thanks!

Comment (nahumsa) [reply pennylane/ops/channel.py:289]: Hi @ixfoduap !  I was reading through the [qiskit's docs](https://qiskit.org/documentation/tutorials/simulators/5_noise_transformation.html) and found a direct reference about the `ResetError` Kraus operators: 
image

This was the main reason that I changed the definition of the channel, do you think that this is reasonable?

Comment (ixfoduap) [reply pennylane/ops/channel.py:289]: Perfect! Thanks for the clarification!

**Review by ixfoduap — APPROVED** [review]
:shipit:
